### PR TITLE
html: parse javascript broken by php

### PR DESCRIPTION
**Pull request by jafl**
The html parser needs to be able to use php as a guest parser.

### CODE REVIEW

**Comment by masatake** [comment]
(I'm working on #3491).

**Comment by masatake** [comment]
Could you rebase this pull request on the latest master?

File: Units/parser-php.r/run-guest-nested.d/args.ctags
Comment (masatake) [inline Units/parser-php.r/run-guest-nested.d/args.ctags:4]: Could you put a new line at the end?

File: Units/parser-php.r/run-guest-nested.d/input.php
Comment (masatake) [inline Units/parser-php.r/run-guest-nested.d/input.php:39]: Could you put a new line at the end?

Comment (masatake) [reply Units/parser-php.r/run-guest-nested.d/input.php:39]: (If possible, I should put a rule to .editorconfig forcing this.)

File: Units/parser-php.r/run-guest-nested.d/args.ctags
Comment (masatake) [inline Units/parser-php.r/run-guest-nested.d/args.ctags:4]: This is not a requirement for merging. Just an idea.

I wonder if we can remove --language-force=html.
If the PHP parser recognizes the input is an HTML file when reading "<HTML", the PHP parser can run the HTML parser as a guest. Then the HTML parser runs PHP parser when reading '<?php'

**Review by masatake — APPROVED** [review]

File: Units/parser-php.r/run-guest-nested.d/args.ctags
Comment (jafl) [reply Units/parser-php.r/run-guest-nested.d/args.ctags:4]: Unfortunately, when PHP is the root parser, it does not find the JavaScript functions.  I don't see any way to fix that in the PHP parser, because it only understands "chunks between PHP code", and cannot treat the entire <script> block as a single unit.

**Review by masatake — COMMENTED** [review]
Can I withdraw my approval?
I would like to understand the relation between this pull request and my old one (https://github.com/masatake/ctags/commit/f476dfe1d31ba7d25d0a1f50f94a756d3fc6cd34) .

File: Units/parser-php.r/run-guest-nested.d/args.ctags
Comment (masatake) [inline Units/parser-php.r/run-guest-nested.d/args.ctags:1]: Do you really want the sorted result?

**Comment by masatake** [comment]
Instead of introducing the new test case, how about extending parser-php.r/run-guest.d/ ?

File: parsers/html.c
Comment (masatake) [inline parsers/html.c:168]: TOKEN_OTHER_SCRIPT is introduced here. However, it seems that this one is not used in this pull request.

Comment (jafl) [reply parsers/html.c:168]: Removed.  It was left over from a previous iteration of the code.

File: Units/parser-php.r/run-guest-nested.d/args.ctags
Comment (jafl) [reply Units/parser-php.r/run-guest-nested.d/args.ctags:1]: It's easier to read.

**Comment by jafl** [comment]
@masatake The difference between this and your previous commit is that there is `<?php ... ?>` in the middle of the JavaScript block.  Without my change, the functions in the JavaScript block are not tagged.

**Review by masatake — APPROVED** [review]

**Comment by masatake** [comment]
Looks good for me.
Thank you.

File: Units/parser-php.r/run-guest-nested.d/args.ctags
Comment (masatake) [reply Units/parser-php.r/run-guest-nested.d/args.ctags]: I inspected the old code that I almost forgot and your change.
In the current implementation, the PHP parser makes a promise for each HTML fragment.
A javascript block is splitted into HTML fragments. As the result, especially the bottom half of the block cannot be parsed well by JavaScript parser.

Instead of making small promises, we can make the PHP parser makes one large promise when "<HTML" is found WITH masking <?php...> and <%...%> areas by whitespace.

A promise can be built with a filter that overwrites the input stream before passing it to a guest parser as far as the filter doesn't change the length of the input stream.
`attachPromiseModifier` defined in promise.c is the engine applying such a filter.
Let me try to apply this algorithm to delete --language-force=html option:-P.

**Comment by masatake** [comment]
Do you like this crazy mechanism for stacking parsers?

More crazy items can be found in https://github.com/universal-ctags/ctags/issues/1577 .
When the JavaScript parser supports destructing binding, I will work on supporting VUE again.

File: Units/parser-php.r/run-guest-nested.d/input.php
Comment (masatake) [reply Units/parser-php.r/run-guest-nested.d/input.php]: (You don't have to take any extra action.)

My comment here is wrong.
About input.*, any contents are acceptable.
The new line at the EOF is generally expected.
However, for the test input, there is no common expectation.
We may want to test ctags with various broken inputs.

Feel free to revert the change you had to introduce for responding to my review comment.

File: Units/parser-php.r/run-guest-nested.d/args.ctags
Comment (masatake) [reply Units/parser-php.r/run-guest-nested.d/args.ctags]: Fypp parser is line-oriented but implements this technique; making a promise for whole the file area to run Fortran parser as a guest.
Line filters attached to the promise mask the preprocessor lines in the area with whitespaces.

**Comment by masatake** [comment]
@jafl don't you merge this?

**Comment by jafl** [comment]
@masatake I thought you were working on a different solution to this issue?

**Comment by masatake** [comment]
> @masatake I thought you were working on a different solution to this issue?

Sorry for making you be confused. I will work on the extra solution after you merge this pull request. So I would like you to merge this pull request if you are o.k.

**Comment by masatake** [comment]
Thank you!

**Pull request merged**
